lib/executor/constant_arrival_rate.go

```diff
@@ -159,44 +165,49 @@ func (carc ConstantArrivalRateConfig) GetExecutionRequirements(es *lib.Execution
 func (carc ConstantArrivalRateConfig) NewExecutor(
 	es *lib.ExecutionState, logger *logrus.Entry,
 ) (lib.Executor, error) {
-	return ConstantArrivalRate{
+	return &ConstantArrivalRate{
 		BaseExecutor: NewBaseExecutor(carc, es, logger),
 		config:       carc,
 	}, nil
 }
 
 // HasWork reports whether there is any work to be done for the given execution segment.
-func (carc ConstantArrivalRateConfig) HasWork(es *lib.ExecutionSegment) bool {
-	return carc.GetMaxVUs(es) > 0
+func (carc ConstantArrivalRateConfig) HasWork(et *lib.ExecutionTuple) bool {
+	return carc.GetMaxVUs(et) > 0
 }
 
 // ConstantArrivalRate tries to execute a specific number of iterations for a
 // specific period.
 type ConstantArrivalRate struct {
 	*BaseExecutor
 	config ConstantArrivalRateConfig
+	et     *lib.ExecutionTuple
 }
 
 // Make sure we implement the lib.Executor interface.
 var _ lib.Executor = &ConstantArrivalRate{}
 
+// Init values needed for the execution
+func (car *ConstantArrivalRate) Init(ctx context.Context) error {
+	car.et = car.BaseExecutor.executionState.ExecutionTuple.GetNewExecutionTupleBasedOnValue(car.config.MaxVUs.Int64)
+	// TODO mvoe the preallocation of VUs here ?
```

Comment:
if by "preallocation" you mean "getting them from the buffer pool in the execution state", then that can't happen here. It needs to be in `Run()`, because the `Init()` function of all executors is called before they all start, which, because of the `startTime` option might not be immediately. Please delete the `TODO` if that was the intention, it has a typo anyway :stuck_out_tongue_closed_eyes:

Reply:
hm ... I forgot about the `startTime`